File: backend/database/ratings.go
package database

func getCohort(ratingSystem RatingSystem, currentRating int) DojoCohort {
	switch ratingSystem {

	case Fide:
	case Ecf:
	case Dwz:
		switch {
		case currentRating < 300:
			return "0-300"
		case currentRating < 400:
			return "300-400"
		case currentRating < 500:
			return "400-500"
		case currentRating < 600:
			return "500-600"
		case currentRating < 700:
			return "600-700"
		case currentRating < 800:
			return "700-800"
		case currentRating < 900:
			return "800-900"
		case currentRating < 1000:
			return "900-1000"
		case currentRating < 1100:
			return "1000-1100"
		case currentRating < 1200:
			return "1100-1200"
		case currentRating < 1300:
			return "1200-1300"
		case currentRating < 1400:
			return "1300-1400"
		case currentRating < 1500:
			return "1400-1500"
		case currentRating < 1600:
			return "1500-1600"
		case currentRating < 1700:
			return "1600-1700"
		case currentRating < 1800:
			return "1700-1800"
		case currentRating < 1900:
			return "1800-1900"
		case currentRating < 2000:
			return "1900-2000"
		case currentRating < 2100:
			return "2000-2100"
		case currentRating < 2200:
			return "2100-2200"
		case currentRating < 2300:
			return "2200-2300"
		case currentRating < 2400:
			return "2300-2400"
		default:
			return "2400+"
		}

	case Chesscom:
		switch {
		case currentRating < 500:
			return "0-300"
		case currentRating < 650:
			return "300-400"
		case currentRating < 750:
			return "400-500"
		case currentRating < 850:
			return "500-600"
		case currentRating < 950:
			return "600-700"
		case currentRating < 1050:
			return "700-800"
		case currentRating < 1150:
			return "800-900"
		case currentRating < 1250:
			return "900-1000"
		case currentRating < 1350:
			return "1000-1100"
		case currentRating < 1450:
			return "1100-1200"
		case currentRating < 1550:
			return "1200-1300"
		case currentRating < 1650:
			return "1300-1400"
		case currentRating < 1750:
			return "1400-1500"
		case currentRating < 1850:
			return "1500-1600"
		case currentRating < 1950:
			return "1600-1700"
		case currentRating < 2050:
			return "1700-1800"
		case currentRating < 2150:
			return "1800-1900"
		case currentRating < 2250:
			return "1900-2000"
		case currentRating < 2350:
			return "2000-2100"
		case currentRating < 2425:
			return "2100-2200"
		case currentRating < 2525:
			return "2200-2300"
		case currentRating < 2600:
			return "2300-2400"
		default:
			return "2400+"
		}

	case Lichess:
		switch {
		case currentRating < 1035:
			return "0-300"
		case currentRating < 1100:
			return "300-400"
		case currentRating < 1165:
			return "400-500"
		case currentRating < 1225:
			return "500-600"
		case currentRating < 1290:
			return "600-700"
		case currentRating < 1350:
			return "700-800"
		case currentRating < 1415:
			return "800-900"
		case currentRating < 1475:
			return "900-1000"
		case currentRating < 1575:
			return "1000-1100"
		case currentRating < 1675:
			return "1100-1200"
		case currentRating < 1750:
			return "1200-1300"
		case currentRating < 1825:
			return "1300-1400"
		case currentRating < 1900:
			return "1400-1500"
		case currentRating < 2000:
			return "1500-1600"
		case currentRating < 2075:
			return "1600-1700"
		case currentRating < 2150:
			return "1700-1800"
		case currentRating < 2225:
			return "1800-1900"
		case currentRating < 2300:
			return "1900-2000"
		case currentRating < 2375:
			return "2000-2100"
		case currentRating < 2450:
			return "2100-2200"
		case currentRating < 2525:
			return "2200-2300"
		case currentRating < 2600:
			return "2300-2400"
		default:
			return "2400+"
		}

	case Uscf:
		switch {
		case currentRating < 350:
			return "0-300"
		case currentRating < 450:
			return "300-400"
		case currentRating < 550:
			return "400-500"
		case currentRating < 650:
			return "500-600"
		case currentRating < 750:
			return "600-700"
		case currentRating < 850:
			return "700-800"
		case currentRating < 950:
			return "800-900"
		case currentRating < 1050:
			return "900-1000"
		case currentRating < 1150:
			return "1000-1100"
		case currentRating < 1250:
			return "1100-1200"
		case currentRating < 1350:
			return "1200-1300"
		case currentRating < 1450:
			return "1300-1400"
		case currentRating < 1550:
			return "1400-1500"
		case currentRating < 1650:
			return "1500-1600"
		case currentRating < 1775:
			return "1600-1700"
		case currentRating < 1875:
			return "1700-1800"
		case currentRating < 1975:
			return "1800-1900"
		case currentRating < 2100:
			return "1900-2000"
		case currentRating < 2200:
			return "2000-2100"
		case currentRating < 2300:
			return "2100-2200"
		case currentRating < 2400:
			return "2200-2300"
		case currentRating < 2500:
			return "2300-2400"
		default:
			return "2400+"
		}

	case Cfc:
		switch {
		case currentRating < 425:
			return "0-300"
		case currentRating < 525:
			return "300-400"
		case currentRating < 625:
			return "400-500"
		case currentRating < 725:
			return "500-600"
		case currentRating < 825:
			return "600-700"
		case currentRating < 925:
			return "700-800"
		case currentRating < 1025:
			return "800-900"
		case currentRating < 1125:
			return "900-1000"
		case currentRating < 1225:
			return "1000-1100"
		case currentRating < 1325:
			return "1100-1200"
		case currentRating < 1425:
			return "1200-1300"
		case currentRating < 1525:
			return "1300-1400"
		case currentRating < 1625:
			return "1400-1500"
		case currentRating < 1725:
			return "1500-1600"
		case currentRating < 1825:
			return "1600-1700"
		case currentRating < 1925:
			return "1700-1800"
		case currentRating < 2025:
			return "1800-1900"
		case currentRating < 2125:
			return "1900-2000"
		case currentRating < 2225:
			return "2000-2100"
		case currentRating < 2325:
			return "2100-2200"
		case currentRating < 2425:
			return "2200-2300"
		case currentRating < 2525:
			return "2300-2400"
		default:
			return "2400+"
		}
	}

	return NoCohort
}

File: common/src/ratings/ratings.ts
import { dojoCohorts, getCohortRangeInt } from '../database/cohort';
import { RatingSystem } from '../database/user';

/**
 * Returns true if the given rating system is a custom system.
 * @param ratingSystem The rating system to check.
 */
export function isCustom(ratingSystem: RatingSystem | string | undefined): boolean {
    return (
        ratingSystem === RatingSystem.Custom ||
        ratingSystem === RatingSystem.Custom2 ||
        ratingSystem === RatingSystem.Custom3
    );
}

export function getRatingBoundary(
    cohort: string,
    ratingSystem: RatingSystem,
    boundaries: Record<string, Record<RatingSystem, number>> = ratingBoundaries,
): number | undefined {
    const cohortBoundaries = boundaries[cohort];
    if (!cohortBoundaries) {
        return undefined;
    }

    const boundary = cohortBoundaries[ratingSystem];
    if (boundary <= 0) {
        return undefined;
    }
    return boundary;
}

/**
 * Returns the minimum rating for the given cohort and rating system.
 * @param cohort The cohort to get the minimum rating for.
 * @param ratingSystem The rating system to get the minimum rating for.
 * @returns The minimum rating for the given cohort and rating system.
 */
export function getMinRatingBoundary(
    cohort: string,
    ratingSystem: RatingSystem,
    boundaries: Record<string, Record<RatingSystem, number>> = ratingBoundaries,
): number {
    const cohortIdx = dojoCohorts.findIndex((c) => c === cohort);
    if (cohortIdx <= 0) {
        return 0;
    }
    return getRatingBoundary(dojoCohorts[cohortIdx - 1], ratingSystem, boundaries) || 0;
}

export function getNormalizedRating(
    rating: number,
    ratingSystem: RatingSystem,
    boundaries: Record<string, Record<RatingSystem, number>> = ratingBoundaries,
): number {
    if (isCustom(ratingSystem)) {
        return -1;
    }

    for (const cohort of dojoCohorts) {
        const x2 = getRatingBoundary(cohort, ratingSystem, boundaries);
        if (!x2) {
            continue;
        }

        if (x2 >= rating) {
            const x1 = getMinRatingBoundary(cohort, ratingSystem, boundaries);

            let [y1, y2] = getCohortRangeInt(cohort);

            if (y1 === -1) {
                y1 = 0;
            }
            if (y2 === -1) {
                y2 = 0;
            }

            const result = ((y2 - y1) / (x2 - x1)) * (rating - x1) + y1;
            return Math.round(result * 10) / 10;
        }
    }

    // We are in the 2400+ cohort if we make it here, so we just extrapolate from the 2300-2400 line
    const x1 = getMinRatingBoundary('2300-2400', ratingSystem, boundaries);
    const x2 = getRatingBoundary('2300-2400', ratingSystem, boundaries) || 0;
    const y1 = 2300;
    const y2 = 2400;
    const result = ((y2 - y1) / (x2 - x1)) * (rating - x1) + y1;
    return Math.round(result * 10) / 10;
}

export const ratingBoundaries: Record<string, Record<RatingSystem, number>> = {
    '0-300': {
        [RatingSystem.Chesscom]: 550,
        [RatingSystem.Lichess]: 1250,
        [RatingSystem.Fide]: 0,
        [RatingSystem.Uscf]: 350,
        [RatingSystem.Ecf]: 400,
        [RatingSystem.Cfc]: 350,
        [RatingSystem.Dwz]: 450,
        [RatingSystem.Acf]: 300,
        [RatingSystem.Knsb]: 400,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '300-400': {
        [RatingSystem.Chesscom]: 650,
        [RatingSystem.Lichess]: 1310,
        [RatingSystem.Fide]: 0,
        [RatingSystem.Uscf]: 460,
        [RatingSystem.Ecf]: 625,
        [RatingSystem.Cfc]: 460,
        [RatingSystem.Dwz]: 540,
        [RatingSystem.Acf]: 395,
        [RatingSystem.Knsb]: 600,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '400-500': {
        [RatingSystem.Chesscom]: 750,
        [RatingSystem.Lichess]: 1370,
        [RatingSystem.Fide]: 0,
        [RatingSystem.Uscf]: 570,
        [RatingSystem.Ecf]: 850,
        [RatingSystem.Cfc]: 570,
        [RatingSystem.Dwz]: 630,
        [RatingSystem.Acf]: 490,
        [RatingSystem.Knsb]: 800,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '500-600': {
        [RatingSystem.Chesscom]: 850,
        [RatingSystem.Lichess]: 1435,
        [RatingSystem.Fide]: 0,
        [RatingSystem.Uscf]: 680,
        [RatingSystem.Ecf]: 1000,
        [RatingSystem.Cfc]: 680,
        [RatingSystem.Dwz]: 725,
        [RatingSystem.Acf]: 585,
        [RatingSystem.Knsb]: 1000,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '600-700': {
        [RatingSystem.Chesscom]: 950,
        [RatingSystem.Lichess]: 1500,
        [RatingSystem.Fide]: 0,
        [RatingSystem.Uscf]: 790,
        [RatingSystem.Ecf]: 1130,
        [RatingSystem.Cfc]: 780,
        [RatingSystem.Dwz]: 815,
        [RatingSystem.Acf]: 680,
        [RatingSystem.Knsb]: 1140,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '700-800': {
        [RatingSystem.Chesscom]: 1050,
        [RatingSystem.Lichess]: 1550,
        [RatingSystem.Fide]: 0,
        [RatingSystem.Uscf]: 900,
        [RatingSystem.Ecf]: 1210,
        [RatingSystem.Cfc]: 880,
        [RatingSystem.Dwz]: 920,
        [RatingSystem.Acf]: 775,
        [RatingSystem.Knsb]: 1280,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '800-900': {
        [RatingSystem.Chesscom]: 1150,
        [RatingSystem.Lichess]: 1600,
        [RatingSystem.Fide]: 0,
        [RatingSystem.Uscf]: 1010,
        [RatingSystem.Ecf]: 1270,
        [RatingSystem.Cfc]: 980,
        [RatingSystem.Dwz]: 1025,
        [RatingSystem.Acf]: 870,
        [RatingSystem.Knsb]: 1400,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '900-1000': {
        [RatingSystem.Fide]: 1450,
        [RatingSystem.Uscf]: 1120,
        [RatingSystem.Chesscom]: 1250,
        [RatingSystem.Lichess]: 1665,
        [RatingSystem.Ecf]: 1325,
        [RatingSystem.Cfc]: 1090,
        [RatingSystem.Dwz]: 1110,
        [RatingSystem.Acf]: 990,
        [RatingSystem.Knsb]: 1450,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1000-1100': {
        [RatingSystem.Fide]: 1500,
        [RatingSystem.Uscf]: 1230,
        [RatingSystem.Chesscom]: 1350,
        [RatingSystem.Lichess]: 1730,
        [RatingSystem.Ecf]: 1390,
        [RatingSystem.Cfc]: 1200,
        [RatingSystem.Dwz]: 1185,
        [RatingSystem.Acf]: 1100,
        [RatingSystem.Knsb]: 1500,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1100-1200': {
        [RatingSystem.Fide]: 1550,
        [RatingSystem.Uscf]: 1330,
        [RatingSystem.Chesscom]: 1450,
        [RatingSystem.Lichess]: 1795,
        [RatingSystem.Ecf]: 1455,
        [RatingSystem.Cfc]: 1300,
        [RatingSystem.Dwz]: 1260,
        [RatingSystem.Acf]: 1210,
        [RatingSystem.Knsb]: 1550,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1200-1300': {
        [RatingSystem.Fide]: 1600,
        [RatingSystem.Uscf]: 1420,
        [RatingSystem.Chesscom]: 1550,
        [RatingSystem.Lichess]: 1850,
        [RatingSystem.Ecf]: 1535,
        [RatingSystem.Cfc]: 1390,
        [RatingSystem.Dwz]: 1335,
        [RatingSystem.Acf]: 1320,
        [RatingSystem.Knsb]: 1600,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1300-1400': {
        [RatingSystem.Fide]: 1650,
        [RatingSystem.Uscf]: 1510,
        [RatingSystem.Chesscom]: 1650,
        [RatingSystem.Lichess]: 1910,
        [RatingSystem.Ecf]: 1595,
        [RatingSystem.Cfc]: 1480,
        [RatingSystem.Dwz]: 1410,
        [RatingSystem.Acf]: 1415,
        [RatingSystem.Knsb]: 1650,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1400-1500': {
        [RatingSystem.Fide]: 1700,
        [RatingSystem.Uscf]: 1600,
        [RatingSystem.Chesscom]: 1750,
        [RatingSystem.Lichess]: 1970,
        [RatingSystem.Ecf]: 1665,
        [RatingSystem.Cfc]: 1570,
        [RatingSystem.Dwz]: 1480,
        [RatingSystem.Acf]: 1510,
        [RatingSystem.Knsb]: 1700,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1500-1600': {
        [RatingSystem.Fide]: 1750,
        [RatingSystem.Uscf]: 1675,
        [RatingSystem.Chesscom]: 1850,
        [RatingSystem.Lichess]: 2030,
        [RatingSystem.Ecf]: 1735,
        [RatingSystem.Cfc]: 1645,
        [RatingSystem.Dwz]: 1560,
        [RatingSystem.Acf]: 1605,
        [RatingSystem.Knsb]: 1750,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1600-1700': {
        [RatingSystem.Fide]: 1800,
        [RatingSystem.Uscf]: 1750,
        [RatingSystem.Chesscom]: 1950,
        [RatingSystem.Lichess]: 2090,
        [RatingSystem.Ecf]: 1805,
        [RatingSystem.Cfc]: 1730,
        [RatingSystem.Dwz]: 1640,
        [RatingSystem.Acf]: 1700,
        [RatingSystem.Knsb]: 1800,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1700-1800': {
        [RatingSystem.Fide]: 1850,
        [RatingSystem.Uscf]: 1825,
        [RatingSystem.Chesscom]: 2050,
        [RatingSystem.Lichess]: 2150,
        [RatingSystem.Ecf]: 1875,
        [RatingSystem.Cfc]: 1825,
        [RatingSystem.Dwz]: 1720,
        [RatingSystem.Acf]: 1790,
        [RatingSystem.Knsb]: 1850,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1800-1900': {
        [RatingSystem.Fide]: 1910,
        [RatingSystem.Uscf]: 1930,
        [RatingSystem.Chesscom]: 2165,
        [RatingSystem.Lichess]: 2225,
        [RatingSystem.Ecf]: 1955,
        [RatingSystem.Cfc]: 1925,
        [RatingSystem.Dwz]: 1815,
        [RatingSystem.Acf]: 1900,
        [RatingSystem.Knsb]: 1910,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '1900-2000': {
        [RatingSystem.Fide]: 2000,
        [RatingSystem.Uscf]: 2055,
        [RatingSystem.Chesscom]: 2275,
        [RatingSystem.Lichess]: 2310,
        [RatingSystem.Ecf]: 2065,
        [RatingSystem.Cfc]: 2060,
        [RatingSystem.Dwz]: 1940,
        [RatingSystem.Acf]: 2000,
        [RatingSystem.Knsb]: 2000,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '2000-2100': {
        [RatingSystem.Fide]: 2100,
        [RatingSystem.Uscf]: 2185,
        [RatingSystem.Chesscom]: 2360,
        [RatingSystem.Lichess]: 2370,
        [RatingSystem.Ecf]: 2165,
        [RatingSystem.Cfc]: 2185,
        [RatingSystem.Dwz]: 2070,
        [RatingSystem.Acf]: 2105,
        [RatingSystem.Knsb]: 2100,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '2100-2200': {
        [RatingSystem.Fide]: 2200,
        [RatingSystem.Uscf]: 2290,
        [RatingSystem.Chesscom]: 2425,
        [RatingSystem.Lichess]: 2410,
        [RatingSystem.Ecf]: 2260,
        [RatingSystem.Cfc]: 2290,
        [RatingSystem.Dwz]: 2185,
        [RatingSystem.Acf]: 2215,
        [RatingSystem.Knsb]: 2200,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '2200-2300': {
        [RatingSystem.Fide]: 2300,
        [RatingSystem.Uscf]: 2395,
        [RatingSystem.Chesscom]: 2485,
        [RatingSystem.Lichess]: 2440,
        [RatingSystem.Ecf]: 2360,
        [RatingSystem.Cfc]: 2395,
        [RatingSystem.Dwz]: 2285,
        [RatingSystem.Acf]: 2330,
        [RatingSystem.Knsb]: 2300,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
    '2300-2400': {
        [RatingSystem.Fide]: 2400,
        [RatingSystem.Uscf]: 2500,
        [RatingSystem.Chesscom]: 2550,
        [RatingSystem.Lichess]: 2470,
        [RatingSystem.Ecf]: 2460,
        [RatingSystem.Cfc]: 2500,
        [RatingSystem.Dwz]: 2385,
        [RatingSystem.Acf]: 2450,
        [RatingSystem.Knsb]: 2400,
        [RatingSystem.Custom]: -1,
        [RatingSystem.Custom2]: -1,
        [RatingSystem.Custom3]: -1,
    },
};
